### PR TITLE
Split DLAF into OBJECT libraries

### PR DESCRIPTION
**Pull request by teonnik**
This PR addresses : https://github.com/eth-cscs/DLA-Future/issues/226 . In particular, (incremental) compile-times for tests, miniapps and future algorithms depending only on `dlaf.core` is significantly improved. 

DLAF is split into object libraries:
- dlaf.core
- dlaf.factorization
- dlaf.solver

The combined library is still under the `DLAF` target.

DLAF_AddTest was refactored a little bit. In particular the dlaf object library is specified via the `LIBRARIES` parameter of the `DLAF_addTest()` function.

Suggestion:
- Move the CACHE varibales from `DLAF_addTest()` to the top CMakeLists.txt file?

Note: The CMake version was bumped to 3.14 as linking object libraries is supported starting 3.13.

EDIT: The PR also adds an option to build DLAF as a shared library. However, there seems to be a problem with `lapackpp`  (the problem is unrelated to this PR, it happens on master too):

```
/usr/bin/ld: /home/teonnik/software/spack/opt/spack/linux-arch-broadwell/gcc-8.4.0/lapackpp-develop-e5l4clzhnoqu5tmrpydlu65owgzy53jo/lib/liblapackpp.a(potrf.cc.o): relocation R_X86_64_PC32 against symbol `_ZTVN6lapack5ErrorE' can not be used when making a shared object; recompile with -fPIC
/usr/bin/ld: final link failed: bad value
clang-10: error: linker command failed with exit code 1 (use -v to see invocation)
make[3]: *** [src/CMakeFiles/DLAF.dir/build.make:270: src/libDLAF.so] Error 1
make[2]: *** [CMakeFiles/Makefile2:1299: src/CMakeFiles/DLAF.dir/all] Error 2
make[1]: *** [CMakeFiles/Makefile2:1306: src/CMakeFiles/DLAF.dir/rule] Error 2
make: *** [Makefile:585: DLAF] Error 2
```

### CODE REVIEW

**Comment by teonnik** [comment]
bors try

**Comment by albestro** [comment]
General question (more a curiosity of mine) about target naming.

When DLAF project is exported, IIRC, CMake exports it in a "namespace" (which is more an unwritten convention) with name `dlaf::<target>`. The current proposal is to have `<target>` to be:
- `dlaf` the all-in-one
- `dlaf.*` for specific ones, e.g. `dlaf.core` 

I'm just wondering if there is any "standard" guideline about naming of these targets. If I'm not wrong about this, with the current proposal the user of the library will have to depend on a cmake target named `dlaf::dlaf` or `dlaf::dlaf.core`. About the former one, I'm pretty sure to have seen it in various projects, I don't have any experience about the latter one.

**Comment by teonnik** [comment]
> I'm just wondering if there is any "standard" guideline about naming of these targets.

I'm not aware of a standard guideline. Using `dlaf::` seems fine to me.
> If I'm not wrong about this, with the current proposal the user of the library will have to depend on a cmake target named dlaf::dlaf or dlaf::dlaf.core

Note that we don't export OBJECT libraries, users will only see the complete library : dlaf::DLAF. I would suggest we keep it that way as external users will not benefit from separate OBJECT libraries.

File: cmake/DLAF_AddTest.cmake [hunk 1]
```diff
@@ -194,10 +147,25 @@ function(DLAF_addTest test_target_name)
     list(APPEND _TEST_ARGUMENTS ${_HPX_EXTRA_ARGS_LIST})
   endif()
 
+  ### Test executable target
+  add_executable(${test_target_name} ${DLAF_AT_SOURCES})
+  target_link_libraries(${test_target_name}
+    PRIVATE
+      ${_gtest_tgt}
+      DLAF_test
```
Comment (albestro) [inline cmake/DLAF_AddTest.cmake:155]: Shouldn't you add something like this?

`$<$<BOOL:${IS_AN_MPI_TEST}>:MPI::MPI_CXX>`

Comment (teonnik) [reply cmake/DLAF_AddTest.cmake:155]: MPI is always linked because of `dlaf.prop`, we shouldn't need this.

**Comment by teonnik** [comment]
I've added an auxiliary object library. Looking forward to feedback.

**Comment by rasolca** [comment]
bors try

**Pull request merged**